### PR TITLE
Stupid Cast Analyzer (with Compiler refactoring)

### PR DESCRIPTION
**Pull request by ddmler**
So i refactored the compiler part of casts out into their own compiler classes (like operators) and the analyzer part into the new stupid cast analyzer (with tests now). Added along the way support for 2 missing casts (array and object), added missing types to the existing casts and fixed a bug with a missing typename in the Type class.

### CODE REVIEW

File: src/Analyzer/Pass/Expression/Casts.php [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+     * @param Expr $expr
+     * @param Context $context
+     * @return bool
+     */
+    public function pass(Expr $expr, Context $context)
+    {
+
+        $castType = CompiledExpression::UNKNOWN;
+
+        switch (true) {
```
Comment (ovr) [inline src/Analyzer/Pass/Expression/Casts.php]: I suggest to use `switch(get_class($expr))`

And `case Expr\Cast\Array_::class` it's more easy to read and can up code quality

Thanks

File: src/Analyzer/Pass/Expression/Casts.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            $context->notice(
+                'stupid.cast',
+                'You are trying to cast \''.$typeName.'\' to \''.$typeName.'\'.',
+                $expr
+            );
+            return true;
+        } elseif ($expr instanceof Expr\Cast\Unset_ && $ExprType === CompiledExpression::NULL) {
+            $context->notice(
+                'stupid.cast',
+                'You are trying to cast \'unset\' to \'null\'.',
```
Comment (ovr) [inline src/Analyzer/Pass/Expression/Casts.php]: For this string I suggest to use

`"You are trying to cast 'unset' to 'null'."`

File: src/Analyzer/Pass/Expression/Casts.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        }
+
+        $compiledExpression = $context->getExpressionCompiler()->compile($expr->expr);
+        $ExprType = $compiledExpression->getType();
+        $typeName = $compiledExpression->getTypeName();
+        
+        if ($castType === $ExprType) {
+            $context->notice(
+                'stupid.cast',
+                'You are trying to cast \''.$typeName.'\' to \''.$typeName.'\'.',
```
Comment (ovr) [inline src/Analyzer/Pass/Expression/Casts.php]: Related to it, `sprintf` will be better

**Comment by ddmler** [comment]
@ovr 
I fixed those things and resolved the merge conflict ... don't know if I did it correct. :grin:

**Comment by ovr** [comment]
@ddmler Can you rebase this branch? I am going to protect `merges` because it brokes commit's history

**Comment by ddmler** [comment]
@ovr rebased :+1:

File: tests/analyze-fixtures/Casts.php [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+[
+    {
+        "type":"stupid.cast",
+        "message":"You are trying to cast 'integer' to 'integer'.",
+        "file":"Casts.php",
+        "line":9
+    },
+    {
+        "type":"stupid.cast",
+        "message":"You are trying to cast 'unset' to 'null'.",
```
Comment (ovr) [inline tests/analyze-fixtures/Casts.php]: Unset to null?)

File: src/Analyzer/Pass/Expression/Casts.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            case Expr\Cast\Object_::class:
+                $castType = CompiledExpression::OBJECT;
+                break;
+            case Expr\Cast\String_::class:
+                $castType = CompiledExpression::STRING;
+                break;
+        }
+
+        $compiledExpression = $context->getExpressionCompiler()->compile($expr->expr);
+        $ExprType = $compiledExpression->getType();
```
Comment (ovr) [inline src/Analyzer/Pass/Expression/Casts.php]: Camel-case style please, first char must be in lower `$ExprType ` -> `$exprType `
Thanks

File: src/Analyzer/Pass/Expression/Casts.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            $context->notice(
+                'stupid.cast',
+                sprintf("You are trying to cast '%s' to '%s'.", $typeName, $typeName),
+                $expr
+            );
+            return true;
+        } elseif (get_class($expr) == Expr\Cast\Unset_::class && $ExprType === CompiledExpression::NULL) {
+            $context->notice(
+                'stupid.cast',
+                "You are trying to cast 'unset' to 'null'.",
```
Comment (ovr) [inline src/Analyzer/Pass/Expression/Casts.php]: Seems It should be "You are trying to cast 'null' to 'unset' (null) ."
Thanks

File: src/Analyzer/Pass/Expression/Casts.php [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+class Casts implements AnalyzerPassInterface
+{
+    /**
+     * @param Expr $expr
+     * @param Context $context
+     * @return bool
+     */
+    public function pass(Expr $expr, Context $context)
+    {
+
```
Comment (ovr) [inline src/Analyzer/Pass/Expression/Casts.php]: Please remove one empty line before `$castType`

**Comment by ovr** [comment]
@ddmler Yeap, you rebased success
Please review small comments and fix related to this commits
After it, this PR will be merged

Thanks

**Comment by ovr** [comment]
LGTM, Thank you for PR again, and again ❤️ 
Rly cool to see new features from you, respect 🙇 

BTW, please for feature commit messages lets write like

"fix things" -> "Small fixes after review"

First letter f -> F

Thanks

**Pull request merged**